man/sequencer64.1
.TH Sequencer64 "May 2021" "Version 0.97.0" "Seq64 Manual Page"

.SH NAME
Sequencer64 - Real-time live MIDI sequencer refactored from Seq24

.SH SYNOPSIS
.B seq64
[\fIOPTIONS\fP] [\fIFILENAME\fP]
.BR
.B seq64cli
[\fIOPTIONS\fP] [\fIFILENAME\fP]
.BR
.B qseq64
[\fIOPTIONS\fP] [\fIFILENAME\fP]

The name of the executable depends on the build: ALSA-only (deprecated), Gtk UI,
command-line, and Qt UI.

.SH DESCRIPTION
.PP
\fISequencer64\fP is a real-time MIDI sequencer. It was created to
provide a simple interface for editing and playing MIDI 'loops'.
It uses gtk-2 or Qt 5 for the user interface.

.SH OPTIONS
Sequencer64 accepts the following options, plus an optional name of
a MIDI file. Note that the legacy options, which use underscores or
are missing the option hyphen, are still unofficially supported.
Note that many of the options are "sticky".  That is, they will
be saved to the "rc" configuration file.

.TP 8
.B  \-h, \-\-help
Display a list of all commandline options.

.TP 8
.B  \-V, \-\-version
Display the program version.

.TP 8
.B  \-v, \-\-verbose
Adds more output to the console, for troubleshooting.  This option
is not saved to the "rc" configuration file.

.TP 8
.B  \-H, \-\-home dir
Set the directory that holds the configuration files.  It is always
relative to the user's HOME directory.  The default is ./config/sequencer64.

.TP 8
.B  \-X, \-\-playlist filename
Load the given play-list from the HOME directory.  The conventional name for
this file is, for example, "live-sets.playlist".
This file provides one or more play-list
entries, each providing a list of one or more songs.  Once loaded, the
user can use the four arrow keys to move between play-lists and the songs
in each play-list.  The play-list entries are also controllable via MIDI
control values set in the "rc" file.  See the Sequencer64 manual for
more information.  Note that, once set, this option is, by default, saved
in the "rc" file.

.TP 8
.B  \-l, \-\-legacy
Save the MIDI file in the old Seq24 format, as unspecified
binary data, instead of as a legal MIDI track with meta events.
Also read the configuration, if provided, from the ~/.seq24rc and ~/.seq24usr
files, instead of the new ~/.config/sequencer64/sequencer64.rc and
~/.config/sequencer64/sequencer64.usr files.  The user-interface will indicate
this mode.  This mode is also used if the application is invoked as the
\fIseq24\fP command (one can create a soft link to the sequencer64 binary to
make that happen).

.TP 8
.B \-b b, \-\-bus b
Supports modifying the buss number on all tracks when a MIDI file
is read.  All tracks are loaded with this buss-number override.  This feature
is useful for testing, making it easy to map the MIDI file onto the system's
current hardware/software synthesizer setup.

.TP 8
.B \-B b, \-\-buss b
The same as the --bus option, just for oldsters to use.

.TP 8
.B \-q p, \-\-ppqn p
Supports modifying the PPQN value of Sequencer64, which defaults
to a value of 192.  This setting should allow MIDI files to play back at the
proper speed, and be written with the new PPQN value.  The support values for
PPQN (p) are 0 and 32 to 19200.  The value of 0 is special.  It indicates
that the Sequencer64 internal PPQN should be set to the value that
is present in the currently-loaded MIDI file.

.TP 8
.B \-L, \-\-lash
If LASH support is compiled into the program, this option
enables it.
If LASH support is not compiled into the program, this option will note
be shown in the output of the --help option.

.TP 8
.B \-n, \-\-no-lash
Disable LASH.  This option is useful if LASH is enabled in the configuration
file, but one wants it disabled.

.TP 8
.B \-\-file \fI<filename>\fP
Load the specified MIDI file on startup.
This option does not exist.
Instead, specify the file itself as the last command-line argument.

.TP 8
.B \-m, \-\-manual-alsa-ports
Sequencer64 won't attach ALSA MIDI ports. Instead it will make "manual"
(virtual) ports.
It will create is own set of virtual input and output busses/ports.
By default, 16 ports (0 to 15) are created.  There is an option in the 'rc' file
(sequencer64.rc) to reduces this number.  The opposite option is '-a'.
Not supported in the PortMIDI build.

.TP 8
.B \-a, \-\-auto-alsa-ports
Sequencer64 will attach ALSA ports.  Useful for overriding the
sequencer64.rc configuration file when set up to use manual (virtual) ports.  The
opposite option is '-m'.

.TP 8
.B \-r, \-\-reveal-alsa-ports
Do not use the 'user' definition for port names, used the actual
port names reported by ALSA. Shows the native system port names.
The opposite option is '-R'.

.TP 8
.B \-R, \-\-hide-alsa-ports
Use the 'user' definition for port names. These option looks up the port names in
the 'user' file, sequencer64.usr, and uses them.  Useful to show the actual
equipment attached to each port.
The opposite option is '-r'.

.TP 8
.B \-A, \-\-alsa
Sequencer64 will use ALSA MIDI ports.  This option is useful to override the
configuration file, if it was set up for JACK.  This is a sticky option.
The opposite option (JACK), is '-t'. JACK transport is also disabled.

.TP 8
.B \-s, \-\-show-midi
Dumps incoming MIDI to screen.

.TP 8
.B \-k, \-\-show-keys
Prints the pressed key value.

.TP 8
.B \-K, \-\-inverse
Changes the sequence editor and performance/song editor colors to an
inverse-color mode.  This mode can be considered a "night mode".

.TP 8
.B \-p, \-\-priority
Runs higher priority with a FIFO scheduler.
This option needs root access to succeed.

.TP 8
.B \-P, \-\-pass-sysex
Passes any incoming SYSEX messages to all outputs.
Not yet fully implemented.

.TP 8
.B \-i, \-\-ignore \fI<number>\fP
Ignore ALSA device <number>.

.TP 8
.B \-d, \-\-record-by-channel
Divert MIDI input by cannel into the sequences that are configured for
each channel.

.TP 8
.B \-D, \-\-legacy-record
Record all MIDI into the active sequence.  Also set by the --legacy
option.

.TP 8
.B \-x, \-\-interaction-method \fI<number>\fP
Select the mouse interaction method
0 = seq24 (the default); and 1 = fruity loops method.
The latter does not completely support all actions supported by the Seq24
interaction method, at this time.

The following options will not be shown by --help if the application is
not compiled for JACK support.

.TP 8
.B \-j, \-\-jack-transport
Sequencer64 will sync to JACK transport.  Note that JACK transport is separate
from native JACK MIDI support.

.TP 8
.B \-J, \-\-jack-master
Sequencer64 will try to be the JACK master.  This is a sticky option.

.TP 8
.B \-C, \-\-jack-master-cond
JACK master will fail if there is already a JACK master.

.TP 8
.B \-M, \-\-jack-start-mode \fI<x>\fP
When Sequencer64 is synced to JACK, the following play modes are available:
0 = live mode and 1 = song mode (the default).  Live mode means that
the musician controls the unmuting of patterns on the main window.  Song mode
means that the song layout in the "Song" window controls the playback.

.TP 8
.B \-t, \-\-jack-midi
Use native JACK MIDI.  This is a separate option from JACK Transport, and is
the default.  If JACK is not running, the application falls back to ALSA.

.TP 8
.B \-N, \-\-no-jack-midi
Use ALSA MIDI, even with JACK Transport.  See the -A option.

.TP 8
.B \-S, \-\-stats
Print statistics on the command-line while running, if built for
showing statistics.

.TP 8
.B \-u, \-\-user-save
Save the "user" configuration file after exiting.  Normally, it is saved
only if it does not exist, so as not to make certain command-line "user"
options (such as --bus) permanent.  (Perhaps the same should be true of the
"rc" configuration options.)

.TP 8
.B \-f, \-\-rc filename
Use a different "rc" configuration file.  It must be a file in the user's
$HOME/.config/sequencer64 directory or the directory specified by the --home
option.  Not supported by the --legacy mode.  The '.rc' extension is added if
no extension is present in the filename.

.TP 8
.B \-F, \-\-usr filename
Use a different "usr" configuration file.  It must be a file in the user's
$HOME/.config/sequencer64 directory or the directory specified by the --home
option.  Not supported by the --legacy mode.  The '.usr' extension is added if
no extension is present in the filename.
.TP 8
.B \-c, \-\-config basename
Use a different configuration file base name for the 'rc' and 'usr' files.
For example, one can specify a full configuration for "testing", for "jack",
or for "alsa".
.TP 8
.B \-o, \-\-option opvalue
Provides additional options, including the no-GUI version of
Sequencer64.  Here are the opvalues supported:

daemonize     Makes this application fork to the background.
no-daemonize  Makes this application not fork to the background.
log=filename  Redirect console output to a log file in the
              --home directory [$HOME/.config/sequencer64].
wid=3x2,f     Sets up for multiple main windows, to show
              multiple sets in a grid of patterns panels.
sets=RxC      Modifies the rows and columns in a set from the
              default of 4x8.  Supported values of R are 4 to 8,
              and C can range from 8 to 12. If not 4x8, seq64 is
              in 'variset' mode. Affects mute groups, too.
scale=x       Scales the main window size, from 0.5 to 3.0.
              A value of 0.75 is useful when using
              "-o wid=2x2 -o sets=8x8", though the pattern
              labelling is mildly distorted.

.SH FILES
\fB$HOME\fP/.config/sequencer64.rc stores the main configuration settings for
Sequencer64.  If it does not exist, it will be generated when Sequencer64
exits.  If it does exist, it will be rewritten with the current configuration
of Sequencer64.  Many, or most, of the command-line options are "sticky", in
that they will be written to the configuration file.

\fB$HOME\fP/.config/sequencer64.usr stores the MIDI-configuration settings and
some of the user-interface settings for Sequencer64.  If it does not
exist, it will be generated with a minimal configuration when Sequencer64
exits.  If it does exist, it will be rewritten with the current configuration
of Sequencer64.  Note that the --legacy option causes the old
configuration-file names to be used.

.SH BUGS
Sequencer64 has them.  See 
https://github.com/ahlstromcj/sequencer64/issues
for the reported bugs.  We take pride in hiding a few more :-D.

.SH SUGGESTIONS AND BUG REPORTS
Any bugs found should be reported to the upstream author and/or package 
maintainer.  See the link in the previous section.

.SH HOMEPAGE
https://github.com/ahlstromcj/sequencer64/

.SH OTHER INFO
--ppqn works, but be aware that it may have bugs.  If a MIDI file is re-saved,
--ppqn is also saved.  If no JACK/LASH options are shown above, they were
disabled in the build configuration.

The current Sequencer64 project homepage is a simple git repository at the
https://github.com/ahlstromcj/sequencer64.git URL.
Up-to-date and more comprehensive instructions can be found in the project at
the https://github.com/ahlstromcj/sequencer64-doc.git URL.

The old Seq24 project homepage is at <http://www.filter24.org/seq24/>, and the
new one is at <https://edge.launchpad.net/seq24/>.  It is released under the
GNU GPL license.  Sequencer64 is also released under the GNU GPL license.

.SH SEE ALSO
There are no man-pages yet for the "rc" and "usr" configuration files.
However, when Sequencer64 is first run, these files are saved in
$HOME/.config/sequencer64, and they are fairly self-documenting.

.SH AUTHOR
Sequencer64 was written by Chris Ahlstrom <ahlstromcj@gmail.com>, with
contributions from Tim Deagan <tim@deagan.net>, Daniel Appelt
<daniel.appelt@gmail.com>, 0rel, layk, and many others.
Seq24 was originally written by Rob C. Buse <seq24@filter24.org> and the
Seq24 team at LaunchPad.

This manual page was written by
Dana Olson <seq24@ubuntustudio.com>
with additions from
Guido Scholz <guido.scholz@bayernline.de>
and
Chris Ahlstrom <ahlstromcj@gmail.com>.



NEWS
NEWS for Sequencer64 0.97.0
Chris Ahlstrom
2015-07-10 to 2021-05-07

This file collects major changes that are older and clutter up
the README file.  This file should be called "OLD NEWS".

Older major updates:

Pull Request #150 from user jfrey-xx:

    While the commit messages should be self-explanatory, here is a list of
    the current modifications:

	-	I encountered weird behaviors while recording without touching the GUI,
        and switching back and forth between play / pause (note on 0, note off
        with some positive values, as if it got inverted at some point) -	I
        am controlling which sequence to record or to set to midi through,
        explicitly disabling for each state the previous sequence when I am
        switching to a new one. Since it appears only one sequence at a time
        (or one sequence/channel) can be an "input" at the level of
        perform.cpp, it could happen that disabling recording or midi through
        "steals" the input from an active sequence.
    -	Incoming MIDI events are systematically checked against all captured
        MIDI, otherwise too often unwanted notes would be recorded or played
        -- in the comments previous behavior is justified to avoid delays, but
        I willing to trade CPU cycles for that.
    -	Fixed toggle for quantizing mode; it was not working neither for MIDI
        control nor GUI.
    -	Biggest change: the MIDI control "reserved for expansion" is now used
        to set the "replace" flag while recording.
    -   Version 0.96.7:
        -   Fixed issue #181 by adding more beats/bar menu values up to 20.
            Note that the fix to issue #77 allows numbers up to 128 to be
            typed in.
        -   In Seq66, found an issue handling running status in the 2rock.mid
            file.  Backported the fix here.
        -   Feature request #188: in the "usr" file, hold the desired state of
            muting, MIDI thru, record, quantized record, and recording style,
            for new patterns when opened.  An assist for live recording.
            Run with the --user-save option to save with the new feature.
            Currently, the criterion for "new pattern" is a name of "Untitled".
        -   Along with #188, made it possible for MIDI control to bring up the
            pattern editor.
        -   Added a work-around for crashing when opening menus with horizontal
            separators on certain themes (e.g. Breeze-Gtk and its dark version).
            See issues #126, #154, and #185.  Currently, all separators are
            removed by default, until we can detect themes.  A slight
            degradation in the quality of the user interface.  Not bad, though.
        -   Added signal handling (e.g. for nsm-proxy) from user falkTX.
            Used in the CLI, GTK, and Qt versions to trigger file save and
            application exit.
        -   Added code to define the engine used and the exact "uname a-" of the
            operating system on which the build was done, for the --version
            option.
    -   Version 0.96.6:
        -   Fixed issue #180 by enabling the (user_ui_) key_height option.
            Change this value in sequencer64.usr from 8 to a larger value as
            desired.
        -   Fixed issue #179 where note lengths recorded were fixed to grid snap
            (in "slave" mode) instead of length played.
        -   Fixed the plain qseqeditframe handling of the play / record /
            qrecord / thru buttons, reported by a Windows user.
        -   Fixed --priority option to use pthread schedule-setting function.
        -   Fixed external seqedit window remaining after closure.
    -   Version 0.96.5:
        -   Merged in the midi-control code of igorangst and fixed some
            misfeatures along the way.
        -   Fixed error in access transpose functionality.
        -   Many minor issues fixed.
    -   Version 0.96.4:
        -   Tweaks back-ported from Sequencer66.
        -   Added arrow movement of notes in qseqroll, and fixed drawing the
            selection box.
        -   Fixed enabling event-editor save button when an event is deleted.
    -   Version 0.96.3:
        A long time in coming, busy with refactoring Seq66 (now available).
        Fixes issue #171 (GitHub), and some minor tweakage.
    -   Version 0.96.2:
        -   Fixed issue with qsliveframe when expanding vertically, and in
            showing hot-keys when an external live-frame with set > 0 is
            created..
        -   Beefing up play-lists to support editing better.
    -   Version 0.96.1:
        -   Fixes to bugs in the Qt live-frame, including coloring.
        -   Varisets implemented in the Qt live-frame.
        -   Fixes to the CLI version of Sequencer64.
        -   Fixes to file import in the Qt version.
        -   Fix to the Qt song-editor transpose combo-box.
        -   File cleanup, usage of PLATFORM_CPP_11 macro.
        -   Fixes to seqkeys issues in Gtkmm and Qt pattern editors.
        -   Fixed GitHub issue #170, where we exited prematurely from
            processing all of the MIDI controls set up in the "rc" file.
        -   Fixed a Qt performance editor seqfault.
        -   Fixed a bug in process the "rc" file [midi-clock] entries.
    -   Version 0.96.0:
        -   Support for playlists and MIDI selections of playlists and songs.
            Includes the command-line option, "rc" file, keystroke selection,
            and MIDI control.
        -   Fixed ALSA issue with port system:announce being enabled in the
            File / Options / MIDI input GUI.
        -   Added an external window for additional Qt live-frame views.
        -   Updated the command-line version, seq64cli, to use the same
            MIDI/WRK file-open function as the rest of the versions.
        -   Fixed the wrkfile parser to set the buss-override value if specified.
        -   Fixed the B:B:T calculation for displaying the current position

Sequencer64/sequencer64.cpp
/*
 *  This file is part of seq24/sequencer64.
 *
 *  seq24 is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  seq24 is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with seq24; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/**
 * \file          sequencer64.cpp
 *
 *  This module declares/defines the main module for the original ALSA-only
 *  implementation of this application.
 *
 * \library       sequencer64 application
 * \author        Seq24 team; modifications by Chris Ahlstrom
 * \date          2015-07-24
 * \updates       2018-04-22
 * \license       GNU GPLv2 or above
 *
 *  Note that there are a number of header files that we don't need to add
 *  here, since other header files include them.
 *
 * \todo
 *      "I don't like seq24's pianoroll editor, the way you do CC envelopes
 *      isn't ideal, it uses alsa-midi, there's unnecessary complexity in
 *      switching from pattern-trigger mode to song mode, and its insistence
 *      on being transport master while not even being able to adjust tempo
 *      when live is annoying."  The JACK support may need updating/upgrading
 *      as well.
 */

#include <stdio.h>
#include <gdkmm/cursor.h>
#include <gtkmm/main.h>

#include "cmdlineopts.hpp"              /* command-line functions           */
#include "daemonize.hpp"                /* seq64::reroute_stdio()           */
#include "file_functions.hpp"           /* seq64::file_accessible()         */
#include "gui_assistant_gtk2.hpp"       /* seq64::gui_assistant_gtk2        */
#include "gui_palette_gtk2.hpp"         /* colors and "inverse" colors      */
#include "lash.hpp"                     /* seq64::lash_driver functions     */
#include "mainwid.hpp"                  /* needed to fulfill mainwnd        */
#include "mainwnd.hpp"                  /* the main window of sequencer64   */
#include "settings.hpp"                 /* seq64::usr() and seq64::rc()     */

/**
 *  The standard C/C++ entry point to this application.  This first thing
 *  this function does is scan the argument vector and strip off all
 *  parameters known to GTK+.
 *
 *  The next thing is to set the various settings defaults, and then try to
 *  read the "user" and "rc" configuration files, in that order.  There are
 *  currently no options to change the names of those files.  If we add
 *  that code, we'll move the parsing code to where the configuration
 *  file-names are changed from the command-line.
 *
 *  The last thing is to override any other settings via the command-line
 *  parameters.
 *
 * \param argc
 *      The number of command-line parameters, including the name of the
 *      application as parameter 0.
 *
 * \param argv
 *      The array of pointers to the command-line parameters.
 *
 * \return
 *      Returns EXIT_SUCCESS (0) or EXIT_FAILURE, depending on the status of
 *      the run.
 */

int
main (int argc, char * argv [])
{
    Gtk::Main kit(argc, argv);              /* strip GTK+ parameters        */
    seq64::rc().set_defaults();             /* start out with normal values */
    seq64::usr().set_defaults();            /* start out with normal values */
    (void) seq64::parse_log_option(argc, argv);    /* -o log=file.ext early */

    /*
     * Set up objects that are specific to the Gtk-2 GUI.  Pass them to the
     * perform constructor.  Then parse any command-line options to see if
     * they might affect what gets read from the 'rc' or 'user' configuration
     * files.  They will be parsed again later so that they can still override
     * whatever other settings were made via the configuration files.
     */

    seq64::gui_assistant_gtk2 gui;              /* GUI-specific objects     */
    seq64::perform p(gui);                      /* main performance object  */
    (void) seq64::parse_command_line_options(p, argc, argv);
    bool is_help = seq64::help_check(argc, argv);
    bool ok = true;
    int optionindex = SEQ64_NULL_OPTION_INDEX;
    if (! is_help)
    {
        /*
         * \todo
         *      If parsing fails at all, we need to report it and disable
         *      usage of the application and saving bad garbage out when
         *      exiting.
         */

        std::string errmessage;                     /* just in case!        */
        ok = seq64::parse_options_files(p, errmessage, argc, argv);
        optionindex = seq64::parse_command_line_options(p, argc, argv);
        if (seq64::parse_o_options(argc, argv))
        {
            /*
             * Process the -o options, also set up in the "usr" file.
             *
             *  Now handled via incrementing optind:    ++optionindex;
             */

            p.seqs_in_set(seq64::usr().seqs_in_set());
            p.max_sets(seq64::usr().max_sets());

            std::string logfile = seq64::usr().option_logfile();
            if (seq64::usr().option_use_logfile() && ! logfile.empty())
                (void) seq64::reroute_stdio(logfile);
        }
        if (ok)
            p.launch(seq64::usr().midi_ppqn());     /* set up performance   */

        if (seq64::usr().inverse_colors())
            seq64::gui_palette_gtk2::load_inverse_palette(true);

        std::string midifname;                      /* start out blank      */
        std::string errmsg = "unspecified error";
        if (ok)
        {
            if (optionindex < argc)                 /* MIDI filename given? */
            {
                std::string fname = argv[optionindex];
                if (seq64::file_accessible(fname))
                {
                    midifname = fname;      // seq24_window.open_file(fname);
                }
                else
                {
                    char temp[256];
                    (void) snprintf
                    (
                        temp, sizeof temp,
                        "? MIDI file not found: %s\n", fname.c_str()
                    );
                    printf(temp);
                    ok = false;
                    errmsg = temp;
                }
            }
        }

        /*
         * Push the mainwnd window onto the stack, with an option for allowing
         * a second perfedit to be created.  Also be sure to pass along the
         * PPQN value, which might be different than the default (192), and
         * affects some of the child objects of mainwnd.
         */

        seq64::mainwnd seq24_window
        (
            p,
            midifname,                  // optional
            seq64::usr().allow_two_perfedits(),
            seq64::usr().midi_ppqn()
#if defined SEQ64_MULTI_MAINWID
            ,
            seq64::usr().block_rows(),
            seq64::usr().block_columns(),
            seq64::usr().block_independent()
#endif
        );

#ifdef CAN_SHOW_MESSAGEBOX
        if (! ok)
            seq24_window.show_message_box(errmsg);
#endif

        if (ok)
        {

#ifdef PLATFORM_LINUX
            if (seq64::rc().lash_support())
                seq64::create_lash_driver(p, argc, argv);
#endif

            kit.run(seq24_window);                  /* run until user quits  */
            p.finish();                             /* tear down performance */
            if (seq64::rc().auto_option_save())
            {
                if (ok)                             /* don't write bad stuff */
                    ok = seq64::write_options_files(p);
            }
            else
                printf("[auto-option-save is off, so not saving config files]\n");

#ifdef PLATFORM_LINUX
            seq64::delete_lash_driver();            /* deletes only if exists */
#endif
        }
        else
            seq24_window.rc_error_dialog(errmessage);
    }
    return ok ? EXIT_SUCCESS : EXIT_FAILURE ;
}

/*
 * sequencer64.cpp
 *
 * vim: sw=4 ts=4 wm=4 et ft=cpp
 */

